[docs] Clarify when to use SparseAdam

janeyx99 opened the pull request:
![image](https://github.com/pytorch/pytorch/assets/31798555/ff19a522-2630-4578-bc0e-6a704aa94d4e)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101465

janeyx99 committed:
[docs] Clarify when to use SparseAdam

[ghstack-poisoned]

janeyx99 commented:
@zabboud Please leave comments on clarity :D

zabboud commented:
> @zabboud Please leave comments on clarity :D

Leaving a comment on the clarity of the documentation for when to use SparseAdam. 

I opened an issue [#101370](https://github.com/pytorch/pytorch/issues/101370) regarding the lack of clarity in the current documentation for SparseAdam on which use cases the current release covers. From the documentation, the following sentence does not tell the user how/when to use SparseAdam

> In this variant, only moments that appear in the gradient get updated, and only those portions of the gradient get applied to the parameters. 

From this sentence, it sounds like SparseAdam should work for sparse-like gradients, but it does not, since the source code checks for the parameters to be dense and the gradients to be sparse, while it does not provide how that use case can be satisfied. Also, there is no sample provided for the specific use cases. 

For more information on this discussion, please see issue [#101370](https://github.com/pytorch/pytorch/issues/101370)

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"




[ghstack-poisoned]

janeyx99 commented:
Haha thanks for the comment of context @zabboud, but I meant to ask if you could read my code update and ensure it is clear to you 😄

zabboud commented on torch/optim/sparse_adam.py:
```diff
@@ -97,11 +97,36 @@ def step(self, closure=None):
+    the optimization by using the sparse layout to do the update computations. This
+    does NOT mean that SparseAdam is especially meant for tensors with a sparse layout,
+    as some of our other optimizers can handle tensors with a sparse layout as well.
+
```
It might be useful to give an example of when this would actually even occur. How one could achieve sparse gradients (where `param.grad.is_sparse` will return `True`) and dense parameters?

janeyx99 replied:
I think I mention it later on--is the later comment sufficient?

janeyx99 replied:
From looking through the embedding implementation in torch (which eventually dispatches to `embedding_sparse_backward`), the sparse tensors are created from keeping track of actual indices and values which get are needed to construct sparse (semantically and in layout) gradients.

zabboud replied:
So instead of declaring a parameter using `nn.Parameter `, I could just declare it using `nn.Embedding(sparse=True)`?

janeyx99 replied:
Oh no, nn.Embedding is an actual module that involves embedding bags and will do computation on your parameters. I don't really know what the implications of using it will be, but it is not a replacement for nn.Parameter.

My earlier answer is suggesting that you may want to rewrite your backward to directly initialize and construct sparse grads, but I don't anticipate this will be easy or trivial. If what you're looking for is a performance win, I would still recommend fused adam!

zabboud commented on torch/optim/sparse_adam.py:
```diff
@@ -97,11 +97,36 @@ def step(self, closure=None):
+    gradient get updated, and only those portions of the gradient get applied to the
+    parameters. This is an optimization, since the 0-values in the gradient will be
+    skipped, and one can presume that this is only an optimization in semantically
+    sparse tensors, meaning there are lots of 0-values that are skipped. We implement
```
> This is an optimization, since the 0-values in the gradient will be skipped, and one can presume that this is only an optimization in semantically **sparse tensors**, meaning there are lots of 0-values that are skipped.

Maybe would help to clarify, is it sparse input tensors that this would apply for?

janeyx99 replied:
No, I mean gradients here too. I'll clarify!

zabboud commented on torch/optim/sparse_adam.py:
```diff
@@ -97,11 +97,36 @@ def step(self, closure=None):
+    advantage of the sparsity of the grads. One example NN module that behaves as such
+    is `nn.Embedding(sparse=True)`.
+
+    If you suspect your gradients are semantically sparse (but do not have sparse layout),
```
it would be useful for users to know that this potentially might be implemented in future releases

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"




[ghstack-poisoned]

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"




[ghstack-poisoned]

albanD commented on torch/optim/sparse_adam.py:
```diff
@@ -97,10 +97,35 @@ def step(self, closure=None):
+
+    This is a semantically different optimizer where only moments that show up in the
+    gradient get updated, and only those portions of the gradient get applied to the
+    parameters. This is an optimization, since the 0-values in the gradient will be
```
```suggestion
    parameters. This is an optimization, since the values not materialized in the sparse Tensor of gradients will be
```

janeyx99 commented on torch/optim/sparse_adam.py:
```diff
@@ -97,10 +97,35 @@ def step(self, closure=None):
+    skipped, and one can presume that this is only an optimization in semantically
+    sparse grads, meaning there are lots of 0-values that are skipped. We implement
+    the optimization by using the sparse layout to do the update computations. This
+    does NOT mean that SparseAdam is especially meant for tensors with a sparse layout,
```
I was trying to say that you don't necessarily have to use SparseAdam when you have sparse gradients, but I think I shall just not say this to prevent further confusion

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"




[ghstack-poisoned]

cpuhrsch commented on torch/optim/sparse_adam.py:
```diff
@@ -97,10 +97,37 @@ def step(self, closure=None):
+    the gradient get applied to the parameters. This is an optimization, since the
+    values not materialized in the sparse Tensor of gradients will be masked out and
+    skipped. Naturally, the more 0s that are masked, the more performant the
+    optimization. We implement this optimization by using sparse layout Tensors to
```
We use sparse layout Tensors to optimize this approximation, which means the more gradients that are masked by being set to 0s, the more performant the optimization.

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"

![image](https://github.com/pytorch/pytorch/assets/31798555/ff19a522-2630-4578-bc0e-6a704aa94d4e)




[ghstack-poisoned]

albanD commented on torch/optim/sparse_adam.py:
```diff
@@ -97,10 +97,47 @@ def step(self, closure=None):
+    3. Apply this mask over the parameters and only apply an update on non-zero values.
+    
+    We use sparse layout Tensors to optimize this approximation, which means the more
+    gradients that are masked by being set to 0s, the more performant the optimization.
```
nit: mention here that part of this optimization is that we detect what is 0 based on the non-materialized value in the sparse Tensor, we don't actually check the value.

albanD reviewed (approved):
Small nit but SGTM otherwise!

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"

![image](https://github.com/pytorch/pytorch/assets/31798555/ff19a522-2630-4578-bc0e-6a704aa94d4e)




[ghstack-poisoned]

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"

![image](https://github.com/pytorch/pytorch/assets/31798555/ff19a522-2630-4578-bc0e-6a704aa94d4e)




[ghstack-poisoned]

zabboud commented on torch/optim/sparse_adam.py:
```diff
@@ -104,14 +104,16 @@ def step(self, closure=None):
-    2. Apply this mask over the running moments and do computation on only the
+    2. f your gradient is [0, 5, 0, 0, 9], the mask would be [0, 1, 0, 0, 1].
+    3. Apply this mask over the running moments and do computation on only the
+
```
Sorry can I ask for clarification here. What you mention in this update, is exactly what I am doing. But the optimizer won't work in this case still, because the gradients themselves need to satisfy `param.grad.is_sparse()`. perhaps this could be added to clarify? Unless I'm missing something?

janeyx99 replied:
The later comments hope to explain that part. I had wanted to motivate the reasoning of the implementation first, but I agree with moving the disclaimer up front.

zabboud replied:
got it ! thanks!

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"

![image](https://github.com/pytorch/pytorch/assets/31798555/ff19a522-2630-4578-bc0e-6a704aa94d4e)




[ghstack-poisoned]

janeyx99 committed:
Update on "[docs] Clarify when to use SparseAdam"

![image](https://github.com/pytorch/pytorch/assets/31798555/ff19a522-2630-4578-bc0e-6a704aa94d4e)




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
